fedimint-client/src/sm/executor.rs

```diff
@@ -390,181 +405,250 @@ impl ExecutorInner {
                 )
                 .await
                 .expect("Autocommit here can't fail");
-
-            None
-        } else {
-            Some(Box::pin(async move {
-                let (first_completed_result, _index, _unused_transitions) =
-                    select_all(transitions).await;
-                first_completed_result
-            }))
         }
+
+        transitions
     }
 
-    async fn execute_next_state_transitions(
+    async fn run_state_machines_executor_inner(
         &self,
-        global_context_gen: &ContextGen,
+        global_context_gen: ContextGen,
+        sm_update_rx: tokio::sync::mpsc::UnboundedReceiver<DynState>,
     ) -> anyhow::Result<()> {
+        // Re-add the same future  receiving on the channel.
+        // Not very elegant, but can't figure out how to make it a quine
+        // and don't know anything more clever.
+        fn re_add_sm_update_receiver(
+            futures: &mut FuturesUnordered<BoxFuture<'static, ExecutorLoopEvent>>,
+            mut sm_update_rx: mpsc::UnboundedReceiver<DynState>,
+        ) {
+            futures.push(Box::pin(async {
+                if let Some(new) = sm_update_rx.recv().await {
+                    ExecutorLoopEvent::New {
+                        state: new,
+                        sm_update_rx,
+                    }
+                } else {
+                    ExecutorLoopEvent::Disconnected
+                }
+            }));
+        }
+
         let active_states = self.get_active_states().await;
-        // TODO: use DB prefix subscription instead of polling
```

Comment:
Note: world doesn't end on db subscription, and when possible it's more efficient to use in-memory channels.

Reply:
It also has a semantic difference. If there are two things that require notifications and sender is responsible for sending notifications, then two sends is the right thing to do.

I'm not a fan of the whole db subscription thing in particular because by definition it couples notifications with database implementation details. Receiver needs to know exact keys (and thus data model) of the sender to communicate. Refactoring of the notifying party can introduce bugs all the same as forgetting to send, especially that we often leave the old keys/variants around for backward compat reasons. The communication channel is implicit: for the reader of the "sender" code it's much harder to figure out the answer to "well, now what, what's going to happen?".

With channels the communication is explicit. "To make the executor notice new state machine, you need to send it a notification, here is that `send`". In practice it's hard to "forget" about it, because it shouldn't be sprinkled all over the code, but in one(few) function doing it, so the error-proneness should not be any worse.

> and required introducing on_commit to even work, which I did in the end

DB notifications are just a special cased `dbtx.on_commit(|| implicit_channel.send(MAGIC_KEY));` with extra steps . :shrug:  *insert change my mind meme here*